research/picai/preprocessing.py

```diff
@@ -0,0 +1,358 @@
+        mri.scans = [resample_img(scan, mri.settings.spacing, is_label=False) for scan in mri.scans]
+
+        # resample annotation to target resolution
+        mri.annotation = resample_img(mri.annotation, mri.settings.spacing, is_label=True)
+
+        return mri
+
+
+class CentreCropAndOrPad(PreprocessingTransform):
+    def __call__(self, mri: Case) -> Case:
+        """
+        Centre crop and/or pad scans and label.
+
+        Args:
+            mri (Case): The case to be processed.
+
+        Returns:
+            Case: The Case after centre crop and/or padding.
+        """
+        mri.annotation = crop_or_pad(mri.annotation, mri.settings.size, mri.settings.physical_size)
+        mri.scans = [
+            crop_or_pad(
+                scan,
+                mri.settings.size,
+                mri.settings.physical_size,
+            )
+            for scan in mri.scans
+        ]
+
+        return mri
+
+
+class AlignOriginAndDirection(PreprocessingTransform):
+    def __call__(self, mri: Case) -> Case:
+        """
+        Align the origin and direction of each scan, and label.
+
+        Args:
+            mri (Case): The case to be processed.
+
+        Returns:
+            Case: The Case after aligning origin and direction.
+        """
+        case_origin, case_direction, case_spacing = None, None, None
+        for scan in mri.scans:
+            # copy metadata of first scan (nnUNet and nnDetection require this to match exactly)
+            if case_origin is None:
+                case_origin = scan.GetOrigin()
+                case_direction = scan.GetDirection()
+                case_spacing = scan.GetSpacing()
+            else:
+                # check if current scan's metadata is almost equal to the first scan
+                np.testing.assert_allclose(scan.GetOrigin(), case_origin)
+                np.testing.assert_allclose(scan.GetDirection(), case_direction)
+                np.testing.assert_allclose(scan.GetSpacing(), case_spacing)
+
+                # copy over first scan's metadata to current scan
+                scan.SetOrigin(case_origin)
+                scan.SetDirection(case_direction)
+                scan.SetSpacing(case_spacing)
+
+        assert case_origin is not None and case_direction is not None and case_spacing is not None
+        mri.annotation.SetOrigin(case_origin)
+        mri.annotation.SetDirection(case_direction)
+        mri.annotation.SetSpacing(case_spacing)
+
+        return mri
+
+
+class BinarizeAnnotation(PreprocessingTransform):
+    def __call__(self, mri: Case) -> Case:
+        """
+        Binarize the granular ISUP ≥ 2 annotations.
+
+        Args:
+            mri (Case): The case to be processed.
+
+        Returns:
+            Case: The Case after binarizing the annotation.
+        """
+        annotation_array = sitk.GetArrayFromImage(mri.annotation)
+
+        # convert granular PI-CAI csPCa annotation to binary csPCa annotation
```

Comment:
Yeah this is for csPCa. The reason we do this is that the annotation are actually on a ISUP scale 1 - 5. Here we consider anything with ISUP greater or equal to 2 as csPCa. (Shifted to 1 below because of 0 indexing)

Reply:
Gotcha!